Switch from schema to clojure.spec

anthonygalea opened the pull request:
This is a first stab at migrating to clojure.spec. I wouldn't call it ready to be merged but thought it would be a good idea to get feedback early.

devth commented:
@anthonygalea looks good so far! I noticed you renamed `yetibot.core.adapters.init` to `yetibot.core.adapters`. I'm fine w/ it but what's the intention? Thanks!

anthonygalea commented:
This is (I think) a minor improvement in naming with the aim of making things more readable. `::config` being a namespaced keyword is `:adapters/config` instead of `:adapters-init/config`. This is both shorter and in line with all the other "configs" ex: `:logging/config`. We don't have `:logging-init/config` even though it is also used for initialization. In addition [here](https://github.com/yetibot/yetibot.core/blob/master/test/yetibot/core/test/adapters/slack.clj#L20) we could have `adapters/config` instead of `ai/adapters-config`.

I typically like to work such changes piecemeal into commits as I encounter them. Having said that if such changes bug you let me know and I'll refrain from doing them.

anthonygalea commented:
Reviewed pull request, made [additional changes](https://github.com/yetibot/yetibot.core/compare/5c77d820cfeebb242d0268c0a3e15683c979635f..216de00f12fa2de4dbafecbe9a0776e312634d36) and rebased on latest master. Also reviewed corresponding [pull request](https://github.com/yetibot/yetibot/pull/959) in yetibot.

devth commented:
Thanks for the explanation! I agree - `:adapters/config` is much nicer. And I'm fine with changes like this - just wanted to understand the intent.

I'm trying to figure out why the build is failing. It looks like the db schemas are empty at https://travis-ci.com/yetibot/yetibot.core/jobs/230115009#L250, so the database isn't being properly setup for test cases, which causes subsequent failures.

anthonygalea commented:
Fixed, it was a bug in the spec for `:schema/spec`.

devth commented on src/yetibot/core/logging.clj:
```diff
@@ -1,17 +1,17 @@
 
 (defn log-level []
-  (let [v (get-config config-schema [:log :level])]
+  (let [v (get-config ::config [:log :level])]
```
Also need to use a spec for the `get-config` call below on line 23: `(get-config String [:log :rolling :enabled])`. 

Otherwise it throws:

```
Execution error (ClassCastException) at yetibot.core.util.config/get-config (config.clj:55).
java.lang.Class cannot be cast to clojure.lang.IFn
```

anthonygalea replied:
Good catch. I think when I tackled this I made the mistake of only looking for schema requires. Now looked for all calls to `get-config` and made sure it always gets a spec.

devth commented on src/yetibot/core/util/config.clj:
```diff
@@ -41,26 +41,23 @@
-           :schema schema}))
+        ;; invalid against spec
+        {:error :invalid
+         :message (str value " does not validate against spec: " (pr-str spec))
```
We could use https://github.com/bhb/expound (or something like it) to provide a better error message here.

anthonygalea replied:
Thanks for suggesting this, never heard of it before. Tried it and I think the output looks good.

anthonygalea commented:
Thanks a lot for your feedback @devth!
Just addressed the comments.

Regarding spec2, thanks a lot for the link explaining the differences. I would prefer to tackle this in a future pull request if it's ok with you.

devth commented:
> Regarding spec2, thanks a lot for the link explaining the differences. I would prefer to tackle this in a future pull request if it's ok with you.

Yep, sounds good!

anthonygalea commented:
Great @devth. Yep good to merge.
